nat/traversal/pinger.go

```diff
@@ -33,69 +35,98 @@ const prefix = "[NATPinger] "
 	pingCancelled  chan struct{}
-	natEventWaiter NatEventWaiter
-	configParser   ConfigParser
 	stop           chan struct{}
 	once           sync.Once
+	natEventWaiter NatEventWaiter
+	configParser   ConfigParser
+	natProxy       natProxy
+	portPool       PortSupplier
+	consumerPort   int
 }
 
 // NatEventWaiter is responsible for waiting for nat events
 type NatEventWaiter interface {
 	WaitForEvent() Event
 }
 
+// ConfigParser is able to parse a config from given raw json
+type ConfigParser interface {
+	Parse(config json.RawMessage) (ip string, port int, serviceType services.ServiceType, err error)
+}
+
+// PortSupplier provides port needed to run a service on
+type PortSupplier interface {
+	Acquire() (port.Port, error)
+}
+
 // NewPingerFactory returns Pinger instance
-func NewPingerFactory(waiter NatEventWaiter, parser ConfigParser) *Pinger {
-	target := make(chan json.RawMessage)
-	received := make(chan struct{})
+func NewPingerFactory(waiter NatEventWaiter, parser ConfigParser, proxy natProxy, portPool PortSupplier) *Pinger {
+	target := make(chan *Params)
 	cancel := make(chan struct{})
 	stop := make(chan struct{})
 	return &Pinger{
 		pingTarget:     target,
-		pingReceived:   received,
 		pingCancelled:  cancel,
+		stop:           stop,
 		natEventWaiter: waiter,
 		configParser:   parser,
-		stop:           stop,
+		natProxy:       proxy,
+		portPool:       portPool,
 	}
 }
 
-// Start starts NAT pinger and waits for pingTarget to ping
+type natProxy interface {
+	handOff(serviceType services.ServiceType, conn *net.UDPConn)
+	registerServicePort(serviceType services.ServiceType, port int)
+	isAvailable(serviceType services.ServiceType) bool
+}
+
+// Params contains session parameters needed to NAT ping remote peer
+type Params struct {
+	RequestConfig json.RawMessage
+	Port          int
+}
+
+// Start starts NAT pinger and waits for PingTarget to ping
 func (p *Pinger) Start() {
 	log.Info(prefix, "Starting a NAT pinger")
+
+	// We dont need to run pinger if NAT port auto-configuration is successful
+	if p.natEventWaiter.WaitForEvent().Type == SuccessEventType {
+		return
+	}
+
 	for {
 		select {
 		case <-p.stop:
 			return
-		case dst := <-p.pingTarget:
-			IP, port, err := p.configParser.Parse(dst)
+		case pingParams := <-p.pingTarget:
+			log.Info(prefix, "Pinging peer with", pingParams)
+
+			IP, port, serviceType, err := p.configParser.Parse(pingParams.RequestConfig)
 			if err != nil {
-				log.Warn(prefix, errors.Wrap(err, fmt.Sprintf("unable to parse ping message: %v", string(dst))))
+				log.Warn(prefix, errors.Wrap(err, fmt.Sprintf("unable to parse ping message: %v", pingParams)))
+			}
+
+			if !p.natProxy.isAvailable(serviceType) {
+				log.Warn(prefix, serviceType, " NATProxy is not available for this transport protocol")
+				continue
 			}
 
-			log.Infof("%s ping target received: IP: %v, port: %v", prefix, IP, port)
+			log.Infof("%sping target received: IP: %v, port: %v", prefix, IP, port)
```

Comment:
prefix missing

Reply:
its not, its format string, prefix at the end.